### PR TITLE
[pytorch][codegen] migrate gen_variable_type to new data model

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49735 [pytorch][codegen] migrate gen_variable_type to new data model**

This is the final wave of autograd codegen data model migration.

After this PR:
- autograd codegen no longer depends on Declarations.yaml;
- autograd codegen sources are fully type annotated and pass mypy-strict check;

To avoid potential merge conflicts with other pending PRs, some structural
changes are intentionally avoided, e.g. didn't move inner methods out, didn't
change all inner methods to avoid reading outer function's variables, and etc.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Confirmed clean mypy-strict run:
```
mypy --config mypy-strict.ini
```

Differential Revision: [D25678879](https://our.internmc.facebook.com/intern/diff/D25678879)

### CODE REVIEW

**Comment by ljk53** [comment]
@albanD I also have it rebased on top of your PR: https://github.com/ljk53/pytorch/commit/313f5cc50b4c2eaa50128bd08d2efe19c998effa

The merge conflicts seem to be manageable - so I think this could go before or after your PR, let me know your plan :)

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -274,203 +275,221 @@
...
+        arg: Union[Argument, SelfArgument, TensorOptionsArguments]
+    ) -> Optional[DifferentiableInput]:
+        if isinstance(arg, TensorOptionsArguments):
+            return None
+        a: Argument = arg.argument if isinstance(arg, SelfArgument) else arg
+
+        # TODO: `cpp_type` is only to keep it byte-for-byte compatible with the old codegen, should remove.
+        # NB: This is not a clone of cpp.argument() - TensorOptionsArguments / faithful / binds are
+        # not handled properly as they are irrelevant for this codegen.
+        cpp_type = cpp.argumenttype_type(a.type, mutable=a.is_write, binds=a.name).cpp_type()
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:423]: nit: `argumenttype_type` is a bit surprising in terms of naming convention :)

File: tools/autograd/gen_variable_type.py [hunk 2]
```diff
@@ -505,75 +524,76 @@ def guard_for(arg: SavedAttribute) -> Optional[str]:
...
+    def emit_check_if_in_complex_autograd_allowlist() -> List[str]:
+        body: List[str] = []
         if base_name in GRADIENT_IMPLEMENTED_FOR_COMPLEX:
             return body
         for arg in differentiable_outputs:
-            name = arg['name']
-            if arg['type'] == 'Tensor' or arg['type'] == 'TensorList':
-                body.append('throw_error_for_complex_autograd({}, "{}");'.format(name, base_name))
+            name = arg.name
+            # TODO: should be `arg.type.is_tensor_like()`?
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:575]: Well I guess an output can only be differentiable if it is tensor like. So, I would expect this check to not do much?

File: tools/autograd/gen_variable_type.py [hunk 3]
```diff
@@ -694,47 +712,48 @@ def wrap_output(return_values, var):
...
+                if not is_tensor_type(return_info.type) and not is_tensor_list_type(return_info.type):
+                    raise RuntimeError(f'{base_name} that return differentiable views can only return Tensor or Tensor[]')
                 # Only allow rebasing of the history if we return a single Tensor
                 # If we are in a no grad block, raise a warning
                 # See NOTE [ View + Inplace detection ] for more details about this logic
-                if return_info['dynamic_type'] == 'TensorList':
+                if is_tensor_list_type(return_info.type):
                     if base_name in MULTI_OUTPUT_SAFE_FUNCTIONS:
-                        creation_meta = "CreationMeta::MULTI_OUTPUT_SAFE"
+                        creation_meta = 'CreationMeta::MULTI_OUTPUT_SAFE'
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:722]: Out of curiosity, is there a difference between `"` and `'` in python? Or is this just for consistency?

File: tools/autograd/gen_variable_type.py [hunk 4]
```diff
@@ -694,47 +712,48 @@ def wrap_output(return_values, var):
...
+            for unpacked_binding in unpacked_bindings:
+                arg = unpacked_binding.name
+                noref_cpp_type = unpacked_binding.ctype.cpp_type(strip_ref=True)
+                if noref_cpp_type == 'TensorList':
                     save_ptrs_stmts += [SAVE_TENSORLIST_STORAGE.substitute(tensorlist_name=arg),
                                         SAVE_TENSORLIST_IMPL.substitute(tensorlist_name=arg)]
                     enforce_same_ptrs_stmts += [ENFORCE_SAME_TENSORLIST_STORAGE.substitute(tensorlist_name=arg),
                                                 ENFORCE_SAME_TENSORLIST_IMPL.substitute(tensorlist_name=arg)]
-                elif simple_type == 'Tensor':
+                elif noref_cpp_type == 'Tensor':
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:756]: We cannot check `is_tensor_like()` here?

File: tools/autograd/gen_variable_type.py [hunk 5]
```diff
@@ -922,62 +928,47 @@ def dispatch_strategy(declaration):
...
-        if declaration['inplace']:
-            assert name.endswith('_')
-            name = name[:-1]
-        elif name.endswith('_out'):
-            name = name[:-4]
-            arguments = [arg for arg in arguments if not arg.get('output', False)]
-    simple_types = ', '.join(arg['simple_type'] for arg in arguments)
-    return f'{name}({simple_types})'
+def is_tensor_type(t: Type) -> bool:
+    # TODO: Should handle optional here?
```
Comment (albanD) [inline tools/autograd/gen_variable_type.py:932]: Given the replacements above, it should not. Otherwise we should add an argument to disable them for most use cases here.

**Review by albanD — COMMENTED** [review]
That looks quite good!
Small comments below.

**Review by ezyang — APPROVED** [review]
Approving to unblock

File: tools/autograd/gen_autograd.py [hunk 6]
```diff
@@ -177,11 +104,11 @@ def is_operator_selected_for_training(decl):
     # Generate VariableType.h/cpp
     if not disable_autograd:
         from .gen_variable_type import gen_variable_type
-        gen_variable_type(out, aten_decls, differentiability_infos, template_path)
+        gen_variable_type(out, native_functions_path, differentiability_infos, template_path, operator_selector)
 
-        from . import gen_trace_type
+        from .gen_trace_type import gen_trace_type
```
Comment (bhosmer) [inline tools/autograd/gen_autograd.py:109]: It's probably worth pulling the inline imports (here and line 106, at least) out to the top if you're back in here before landing.

File: tools/autograd/gen_variable_type.py [hunk 7]
```diff
@@ -274,203 +275,221 @@
...
-        if len(returns) == 1:
-            return returns[0]['name']
-        moved = ['std::move({})'.format(r['name']) for r in returns]
-        return 'std::make_tuple({})'.format(', '.join(moved))
-
-    return (declare_returned_variables(), tie_return_values(), get_return_value())
+@dataclass(frozen=True)
+class NativeFunctionWithDifferentiabilityInfo:
+    func: NativeFunction
+    info: Optional[DifferentiabilityInfo]
```
Comment (bhosmer) [inline tools/autograd/gen_variable_type.py:281]: note to self: understand why this needs to be optional

File: tools/autograd/gen_variable_type.py [hunk 8]
```diff
@@ -922,62 +928,47 @@ def dispatch_strategy(declaration):
...
-        declaration['non_differentiable_arg_names'] = info.non_differentiable_arg_names if info else []
-        declaration['output_differentiability'] = info.output_differentiability if info else None
+        return functional_info_by_signature.get(f.func.signature()), False
+
+    result: List[NativeFunctionWithDifferentiabilityInfo] = []
+    for f in native_functions:
+        info, is_exact_match = find_info(f)
+        result.append(NativeFunctionWithDifferentiabilityInfo(
+            func=f,
+            info=info,
```
Comment (bhosmer) [inline tools/autograd/gen_variable_type.py:971]: I could definitely be missing something obvious, but when might a `None` creep in here (necessitating its optionality)?

File: tools/jit/gen_unboxing_wrappers.py [hunk 9]
```diff
@@ -279,6 +280,66 @@ def argument_order(decl):
...
+
+    opt_match = re.match(r'c10::optional<(.+)>', simple_type)
+    if opt_match:
+        simple_type = '{}?'.format(opt_match.group(1))
+    return simple_type
+
+
+def load_aten_declarations(path):
+    with open(path, 'r') as f:
+        declarations = yaml.load(f, Loader=YamlLoader)
```
[the remark this answers is not in the document]
Comment (bhosmer) [reply tools/jit/gen_unboxing_wrappers.py:306]: @albanD JFYI after this PR the mainline autograd codegen no longer uses this at all, and after Sebastian's c10-full stuff it should be gone completely (along with `gen_unboxing_wrappers.py`)

**Review by bhosmer — APPROVED** [review]
Hey, looks good! Couple of very minor things noted inline, but not worth holding up landing, only if you're back in for some other reason.

(Though I don't know enough to know if Alban's question on `DifferentiableOutput` implies an issue.)

File: tools/jit/gen_unboxing_wrappers.py [hunk 10]
```diff
@@ -279,6 +280,66 @@ def argument_order(decl):
...
 
 
+def format_return_type(returns):
+    if len(returns) == 0:
+        return 'void'
+    elif len(returns) == 1:
+        return returns[0]['type']
+    else:
+        return_types = [r['type'] for r in returns]
+        return 'std::tuple<{}>'.format(','.join(return_types))
```
[the remark this answers is not in the document]
Comment (ljk53) [reply tools/jit/gen_unboxing_wrappers.py:290]: This entire file is going to be deprecated and the modifications are all simple copy-paste.
I moved old flavor code from gen_autograd to here so all autograd code can pass mypy-strict check.

File: tools/autograd/gen_autograd.py (hunk 6 above)
Comment (ezyang) [reply tools/autograd/gen_autograd.py]: fwiw, the codegen is still simple enough and free of BC concerns that we should be able to come up with some organization that doesn't have circular problems. But I agree, it doesn't matter for codegen, inline import or whatever you want to fix the problem is fine.

File: tools/autograd/gen_variable_type.py (hunk 7 above)
Comment (bhosmer) [reply tools/autograd/gen_variable_type.py:311]: Ah sure - might be worth pasting your answer in here as a comment, if you do another edit to this file.

File: tools/codegen/api/autograd.py [hunk 11]
```diff
@@ -87,3 +87,36 @@ class DifferentiabilityInfo:
...
+# - It can represent SelfArgument or regular Argument but not TensorOptionsArgument;
+@dataclass(frozen=True)
+class DifferentiableInput:
+    name: str
+    type: Type
+
+    # TODO: only to keep it byte-for-byte compatible with the old codegen, should remove.
+    cpp_type: str
+
+# Represents a differentiable `Return`.
```
[the remark this answers is not in the document]
Comment (albanD) [reply tools/codegen/api/autograd.py:108]: My question is just about the comment.
But in here, when there is a `mul_` in native_functions but only `mul` defined in derivatives, we re-use the formula and do things like replacing "result" by "self".
It is just to try and capture in the comment the exact state that is saved here: before or after the adaptation based on the exact variant we're working with.

File: tools/codegen/api/types.py [hunk 12]
```diff
@@ -31,14 +31,16 @@ class BaseCType:
...
         return self.type
 
 @dataclass(frozen=True)
 class ConstRefCType:
     elem: 'CType'
 
-    def cpp_type(self) -> str:
+    def cpp_type(self, *, strip_ref: bool = False) -> str:
+        if strip_ref:
+            return self.elem.cpp_type(strip_ref=strip_ref)
```
[the remark this answers is not in the document]
Comment (albanD) [reply tools/codegen/api/types.py:43]: I just find the structure of the function hard to follow. Maybe an `else` statement would make it clearer?
But this is just a style thing, I don't think this the code is wrong.
But something like this is more natural to me:
```python
no_ref_type = self.elem.cpp_type()
if strip_ref:
  return no_ref_type
else:
  return f'const {no_ref_type} &'
```

File: tools/jit/gen_unboxing_wrappers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (albanD) [reply tools/jit/gen_unboxing_wrappers.py:290]: This was not related to mypy actually. I meant in the generated string so that the c++ codegened code is easier to read

File: tools/autograd/gen_autograd.py (hunk 6 above)
Comment (bhosmer) [reply tools/autograd/gen_autograd.py]: > Curious - what's the problem with inline import? I remember you mentioned we discovered some performance issue with inline import, but why it matters for codegen?

Oh definitely not a perf concern here (or most places), just hygiene.

File: tools/autograd/gen_variable_type.py [hunk 13]
```diff
@@ -959,62 +970,47 @@ def dispatch_strategy(declaration):
...
         # i.e mul() for mul_() or mul_out()
-        signature = get_decl_signature(declaration, use_base_variant=True)
-        return info_by_signature.get(signature)
-
-    for declaration in declarations:
-        info = find_info(declaration)
-        declaration['derivative'] = info if info and info.args_with_derivatives else None
-
-        # Currently, the '.strides()' to 'strides_or_error' replacement does not support
-        # 'self' derivatives of an inplace function, so we must check for this case.
```
Comment (ljk53) [inline tools/autograd/gen_variable_type.py:1010]: oops - I think I missed this check while solving merge conflicts... sorry!

I had a revision with your PR fully patched which contains this check: https://github.com/ljk53/pytorch/commit/313f5cc50b4c2eaa50128bd08d2efe19c998effa#diff-f6553cec68caeaea36f6c8b14ff76a6d39dfd774e0ea9ef2f76e8d81fd9af5dfR1096-R1105

@albanD I can send out a PR to add this check back, or you can help fix it if you are about to land your final codegen PR - please let me know :)

Comment (albanD) [reply tools/autograd/gen_variable_type.py:1010]: I can add that back sure. Thanks!
